lib/vfs.js

```diff
@@ -1,65 +1,93 @@
+            var action = "getting file with path " + path;
+
+            request(url, function (err, response, body) {
+                err = err || utils.resolveHttpError(response, "Error " + action + ".");
+
+                if (err) {
+                    return cb(err);
+                }
+
                 cb(null, body, response);
             });
         },
+
         listFiles: function listFiles(path, cb) {
-            if ('/' !== path[path.length-1]) {
-                path += '/';
-            }
-            request("/api/vfs/" + path, function(err, response, body) {
-                if (err) return cb(err);
-                cb(null, JSON.parse(body), response);
-            });
+            path = ensureDirectory(path);
+
+            var url = "/api/vfs/" + path;
+            var action = "listing files with path " + path;
+
+            request(url, utils.createJsonCallback(cb, action));
         },
+
         uploadFile: function uploadFile(localPath, destPath, etag, cb) {
             if (typeof etag === "function") {
                 cb = etag;
                 etag = "*";
             }
-            fs.createReadStream(localPath).pipe(request.put({
+
+            var options = {
                 uri: "/api/vfs/" + destPath,
                 headers: {
                     "If-Match": etag
                 }
-            }, function(err, response, body) {
-                if (err) return cb(err);
-
-                if (response.statusCode >= 400) {
-                    cb(JSON.parse(body), response);
-                }
+            };
+            var action = "uploading file to destination " + destPath;
 
-                cb(null, response);
-            }));
+            fs.createReadStream(localPath)
+                .pipe(request.put(options, utils.createCallback(cb, action)));
         },
+
         createDirectory: function createDirectory(path, cb) {
-            if ('/' !== path[path.length-1]) {
-                path += '/';
-            }
-            request.put("/api/vfs/" + path, function(err, response, body) {
-                if (err) return cb(err);
-                cb(null, body, response);
-            });
+            path = ensureDirectory(path);
+
+            var url = "/api/vfs/" + path;
+            var action = "creating directory with path " + path;
+
+            request.put(url, utils.createCallback(cb, action));
         },
-        deleteFile: function deleteFile(path, cb) {
-            request.del("/api/vfs/" + path, function(err, response, body) {
-                if (err) return cb(err);
-                cb(null, body, response);
-            });
+
+        deleteFile: function deleteFile(path, etag, cb) {
+            if (typeof etag === "function") {
+                cb = etag;
+                etag = "*";
+            }
+
+            var options = {
+                uri: "/api/vfs/" + path,
+                headers: {
+                    "If-Match": etag
+                }
```

Comment:
Because the old signature is still valid, this isn't a breaking change. I might cherry pick it into a new branch so it can go ahead of the bigger changes in this pull request.